### ISSUE TITLE
Multi-architecture support

### ISSUE THREAD

**Issue by samip5** [phusion/passenger-docker#286]
Has anyone thought about creating multi-arch builds of this image?
There's only amd64 as tagged release, why? aarch64 should be there too, currently this cannot be used in Raspberry Pi, running on arm64 aka aarch64 nor armhf.

**Comment by CamJN** [comment]
The way the docker image works we’d need to serve arm packages from our apt repo, to get that to happen you’d need to add support to our apt automation repo.

**Comment by Theaxiom** [comment]
@CamJN We have multi-arch releases for https://github.com/phusion/baseimage-docker -- I'm wondering if that changes things?

FYI: Just released new version of baseimage (bionic-1.0.0) and hoping to release `focal-1.0.0` very soon, after we thoroughly test it.

**Comment by kurko** [comment]
With the release of M1 Macs, this is a bigger pain point now.

**Comment by CamJN** [comment]
The fastest way to make this happen is filing bugs or better PRs against the https://github.com/phusion/passenger_apt_automation repo to get arm builds working there. Once that's done, updating the passenger docker image is pretty easy.

**Comment by samip5** [comment]
> The fastest way to make this happen is filing bugs or better PRs against the https://github.com/phusion/passenger_apt_automation repo to get arm builds working there. Once that's done, updating the passenger docker image is pretty easy.

Why hasn't https://github.com/phusion/passenger_apt_automation/pull/34/ been merged?

**Comment by CamJN** [comment]
I was a bit unsure about the default being that x86 and x86_64 not needing cross compile, yeah. It's probably better to detect the current architecture and then determine the need for cross compilation from there. I was planning on testing it when i could but I don't have an arm box to test on so it was waiting for that to happen.

**Comment by Theaxiom** [comment]
@CamJN you can use a RaspberryPi :)

**Comment by tiredenzo** [comment]
Hi everyone

It would indeed be nice to have arm64 support for `passenger-docker`, so far building the image manually fails on installing `libnginx-mod-http-passenger`.


I tried building using the following command


```
rm -rf ruby26_image
cp -pR image ruby26_image
echo ruby26=1 >> ruby26_image/buildconfig
echo final=1 >> ruby26_image/buildconfig
docker buildx build -t phusion/passenger-ruby26:2.2.0 --rm ruby26_image --no-cache --platform linux/arm64/v8
```

Thanks in advance

**Comment by tiredenzo** [comment]
On a side note passenger base image runs on ubuntu 20:04, it will be faulty on ARM machines due to a known error in libssl

https://docs.docker.com/desktop/mac/troubleshoot/#known-issues

They recommend updating to ubuntu 21 to solve this issue.

**Comment by badaldavda8** [comment]
> On a side note passenger base image runs on ubuntu 20:04, it will be faulty on ARM machines due to a known error in libssl (segfault)
> 
> https://docs.docker.com/desktop/mac/troubleshoot/#known-issues
> 
> They recommend updating to ubuntu 21 to solve this issue, this version is not an LTS 😕

Ubuntu 22 now available as LTS. We should consider adding arm64 images now. With the release of Graviton processors on aws multiple customers would want to leverage better performance to price ratio. Can anyone from the team please attempt this now?

**Comment by swikrit** [comment]
> With the release of M1 Macs, this is a bigger pain point now.

Could this be related to my issue where I get the this when I try to run this on my M1? 

```
[1] development-passenger-1  | 2022/04/28 10:17:53 [error] 148#148: failed to initialize Lua VM in /etc/nginx/nginx.conf:66
```

None of us had even heard of Lua VM before this. I still have my old machine for when I need to run passenger but it's not ideal. I ran out of keywords to search for and blue links on google to click on 😫

**Comment by samip5** [comment]
> > With the release of M1 Macs, this is a bigger pain point now.
> 
> Could this be related to my issue where I get the this when I try to run this on my M1?
> 
> ```
> [1] development-passenger-1  | 2022/04/28 10:17:53 [error] 148#148: failed to initialize Lua VM in /etc/nginx/nginx.conf:66
> ```
> 
> None of us had even heard of Lua VM before this. I still have my old machine for when I need to run passenger but it's not ideal. I ran out of keywords to search for and blue links on google to click on 😫

Do you have rosetta 2 enabled? It may help or not, but that's an weird error to even have.

**Comment by swikrit** [comment]
> > > With the release of M1 Macs, this is a bigger pain point now.
> > 
> > 
> > Could this be related to my issue where I get the this when I try to run this on my M1?
> > ```
> > [1] development-passenger-1  | 2022/04/28 10:17:53 [error] 148#148: failed to initialize Lua VM in /etc/nginx/nginx.conf:66
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > None of us had even heard of Lua VM before this. I still have my old machine for when I need to run passenger but it's not ideal. I ran out of keywords to search for and blue links on google to click on 😫
> 
> Do you have rosetta 2 enabled? It may help or not, but that's an weird error to even have.

I have rosetta 2 installed, but I'm not sure if it's "enabled". I couldn't find anything on how to enable it for docker, I assumed it happened automatically

**Comment by CamJN** [comment]
rosetta won't be involved with docker for mac except possibly for the electron UI (there is an apple silicon version of the application available to run even the UI without rosetta). 

Docker uses qemu to emulate cross-arch VMs in which it runs the containers.

The lua VM message is probably due to the nginx lua module having been loaded, which image are you using?

**Comment by CamJN** [comment]
which tag? latest? the lua module isn't installed by default:

`# fgrep -RHie lua /etc/nginx/`

**Comment by swikrit** [comment]
Yeah the latest tag. 

We've never had to install the lua module on other devices, is it something specific to the M1 chip?

I'll tried to install Lua on the passenger container (via sudo and also ran the fgrep command) when building but it didn't work. I'm probably trying to install it wrong though, I'm a fair bit over my head with this setup stuff.

**Comment by CamJN** [comment]
No, you shouldn't need lua unless you are using it (you aren't so no). And it's almost certainly not ARM related.

**Comment by badaldavda8** [comment]
Any update on this?

**Comment by dil-bparihar** [comment]
> https://docs.docker.com/desktop/mac/troubleshoot/#known-issues

i am also facing the same issue on my m1 pro laptop

**Comment by badaldavda8** [comment]
> Any update on this?

Any update on this?

**Comment by samip5** [comment]
> > Any update on this?
> 
> Any update on this?

This still waits for the support for arm on the automation repo to my knowledge which has seen no traction.

**Comment by taleh007** [comment]
@CamJN Could you please describe what we need to do in public repositories to add cross-platform images?
As for me it looks like we need to do the following
1. Build `passenger-enterprise`, `libnginx-mod-http-passenger-enterprise`, `nginx`, `passenger`,  `libnginx-mod-http-passenger` for arm and put it into [repository](https://github.com/phusion/passenger-docker/blob/892508dd85e9007ba826f3d49f9a8ef2a191c8f1/image/enable_repos.sh#L9)
2. Migrate to buildx  into makefiles.

Anything missing? Because for now I need to manually build packages from step1 and build images for arm to make it works in my project. And it's pretty fine for me in development but I wanna to help (if needed) to make it public available.

Thanks in advance

**Comment by CamJN** [comment]
If on an apple silicon mac, create a real linux vm, docker for mac cannot be used due to limitations in its filesystem driver.
The actual work to getting this added is fixing the issues you find along the way, or reporting if there were none (unlikely).
The steps to build and test arm images are as follows:

- checkout the passenger code base: `git clone https://github.com/phusion/passenger.git`
- initialize the submodules `cd passenger && git submodule update --init --recursive`
- go to apt automation repo: `cd packaging/debian`
- build the buildbox & testboxes for arm: `cd docker-images && make all`
- run the build script: `cd ../ && ./build -a arm64 -t -p ../../ -c /tmp/c -o /tmp/o -w /tmp/w`
- run the test script `for distro in xenial bionic focal jammy stretch buster bullseye ; do ./test -D -p ../../ -d "/tmp/o/$distro" -c /tmp/c -x "$distro"`

Then once that's working make a PR, that will enable us to put arm packages in our apt repo. After which building arm docker images is trivial (just an extra flag I'll add to the build command in the makefile).

**Comment by taleh007** [comment]
@CamJN done.
I face only with two small problems during build.
1. `The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file` it was fixed by updating `buildbox` to the latest `phusion/baseimage:jammy-1.0.1`
2. Constant link to amd64 files inside tests, also fixed by using current Arch instead.

All tests green and waiting your review

**Comment by CamJN** [comment]
Thanks, i'm digging into this.

**Comment by samip5** [comment]
> @CamJN done. I faced only with two small problems during build.
> 
>     1. `The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file` it was fixed by updating `buildbox` to the latest `phusion/baseimage:jammy-1.0.1`
> 
>     2. Constant link to amd64 files inside tests, also fixed by using current Arch instead.
> 
> 
> All tests green and waiting your review

There's a newer baseimage for jammy available.

**Comment by taleh007** [comment]
> > @CamJN done. I faced only with two small problems during build.
> > ```
> > 1. `The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file` it was fixed by updating `buildbox` to the latest `phusion/baseimage:jammy-1.0.1`
> > 
> > 2. Constant link to amd64 files inside tests, also fixed by using current Arch instead.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > All tests green and waiting your review
> 
> There's a newer baseimage for jammy available.

I dont think so


![image](https://user-images.githubusercontent.com/12513631/192094889-421a1009-9470-4c16-ae14-e0057bdf713b.png)

**Comment by samip5** [comment]
> > > @CamJN done. I faced only with two small problems during build.
> > > ```
> > > 1. `The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file` it was fixed by updating `buildbox` to the latest `phusion/baseimage:jammy-1.0.1`
> > > 
> > > 2. Constant link to amd64 files inside tests, also fixed by using current Arch instead.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > All tests green and waiting your review
> > 
> > 
> > There's a newer baseimage for jammy available.
> 
> I dont think so
> 
> ![image](https://user-images.githubusercontent.com/12513631/192094889-421a1009-9470-4c16-ae14-e0057bdf713b.png)

Forgot it was a dev one temporaily only.

**Comment by taleh007** [comment]
@CamJN Anything else I can do to help you add this?

**Comment by CamJN** [comment]
@taleh007 I don't think so, i'm working on it but also have other things i've got to do, I think I can get a release with arm support made soon now that the work is done and tested, i'll just need to find a few contiguous hours. I'll probably make the first arm images before I get the apt repo updated to host them, since that's more work that no-one can help me with, but since all the docker images are made on my laptop anyway, it's not really much different than how things usually go.

**Comment by taleh007** [comment]
Got it, no rush from side to you I totally understand that it's not one thing you working on. I thought it can be anything that you can delegate to make it more easier. Anyway, Im ready to help if needed

**Comment by CamJN** [comment]
If you want to help out, I've hit a snag with building the packages. The `build` script in the apt automation repo is triggering the following error:
```
pkg:env:focal:arm64           | 2022-09-28 17:06:28:     + pbuilder-dist focal arm64 create --updates-only
pkg:env:focal:arm64           | 2022-09-28 17:06:32:     sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?
pkg:env:focal:arm64           | 2022-09-28 17:06:32: *** Command failed: ./setup-environment.sh focal arm64
```

I've checked and the fs of the docker host does have setuid enabled, so the simplest explanation isn't the cause. I'd appreciate thoughts, or a recommendation of what host OS you built the packages on so I can try that (I used ubuntu which might have some clever setting enabled that causes this error).

**Comment by sjf-control** [comment]
> @CamJN done. I faced only with two small problems during build.
> 
> 1. `The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file` it was fixed by updating `buildbox` to the latest `phusion/baseimage:jammy-1.0.1`

@taleh007 - can you explain what you did to implement step 1 above?
When I execute "make all" I get four errors... 

> E: The repository 'http://ports.ubuntu.com/ubuntu-ports impish Release' does not have a Release file.
> E: The repository 'http://ports.ubuntu.com/ubuntu-ports impish-updates Release' does not have a Release file.
> E: The repository 'http://ports.ubuntu.com/ubuntu-ports impish-backports Release' does not have a Release file.
> E: The repository 'http://ports.ubuntu.com/ubuntu-ports impish-security Release' does not have a Release file.

But I'm not sure how to change the buildbox version.

Thanks!

**Comment by CamJN** [comment]
@sjf-control the latest commits in the passenger apt automation repo have these changes applied, if you cd to  `packaging/debian` and `checkout master && git pull -r` that should help.

**Comment by sjf-control** [comment]
@CamJN is your error showing up in the 'build all' step, or the 'build -a arm64' step?

Your help above got me to the 'build all' step, but all four of the configurations that I've tried have errored out (in different places)

Mac Studio (M1 chip) running VM ubuntu aarch64
Mac Studio (M1 chip) running VM ubuntu x86-64
iMac (Intel chip) running VM ubuntu x86-64
Raspberry Pi-4  running raspios-arm64-lite (no VM)

**Comment by CamJN** [comment]
i'm seeing the error after running `./build -p ~/passenger/ -w ~/w/ -c ~/c/ -o ~/o/ -j 2 -d focal -a arm64 pkg:all` I've already built my buildbox image (and manually tagged it to `phusion/passenger_apt_automation_buildbox:2.0.5` instead of publishing a manifest because i'm not yet ready to publish the images on docker hub because this is all really experimental atm).

**Comment by taleh007** [comment]
> I'd appreciate thoughts, or a recommendation of what host OS you built the packages on so I can try that


I used AWC EC2 instance with next setup:
1. Ubuntu Server 22.04 LTS (HVM), SSD Volume Type
2. Type: c6g.4xlarge

**Comment by taleh007** [comment]
Im trying right now to run `make all` with latest code (where you add both platforms as separated). And it takes a lot of time just build `docker buildx build --platform linux/amd64 --pull --force-rm -t phusion/passenger_apt_automation_buildbox:2.0.5-amd64` on ARM. It takes already 1158.3s and still not finishes. I want to understand what setup you use or plan to use for building this? Is it CI, cloud instance (EC2, digitalocean or etc), local computer?  I think it will require 2 separate machines for building ARM64 and AMD64.

**Comment by CamJN** [comment]
The normal way of things is:

The build & test docker images are built on my laptop
The built images are pushed to docker hub
The CI pulls those images to build packages
The packages are pushed to our repo
The passenger docker image is built on my laptop
The passenger docker image is pushed to docker hub


This time in order to test things out, I can't build the packages in our CI since there's no release going on and I'm not sure that the setup is bug-free yet, so i'm building the docker images and packages in an ubuntu 22.04 vm. 

In my experience building the cross-compiled images took about 4000s each, you could probably edit the makefile to not build the amd64 image since we don't need it for testing.

**Comment by taleh007** [comment]
Also I run `./test -p ~/passenger/ -x focal -D -c ~/c -d ~/o/focal`

And all fine
Log below
[Uploading test_arm64_focal.log…]()

**Comment by sjf-control** [comment]
@taleh007 your link to the log doesn’t seem to work. Just comes back to here.

**Comment by taleh007** [comment]
it should work now

**Comment by taleh007** [comment]
@CamJN Any news? As I mention above (with logs) all looks fine for me if you run builds on ARM based machine.  What else I can do to help?

**Comment by CamJN** [comment]
I've asked for an arm server, we'll see. I was hoping to be able to do this using docker's qemu emulation.

**Comment by CamJN** [comment]
I'm pushing some arm images to docker hub, they'll be tagged like the usual ones but with an `-arm64` suffix. Try them out and let me know how it goes. If all is well I'll finish up including these changes into our CI and Passenger 6.0.16 should have simultaneous arm64 and amd64 images published (and I'll try and make the nice manifest work so you don't need the `-arm64` suffix next time).

**Comment by CamJN** [comment]
no it would be `phusion/passenger-full:2.3.1-arm64` for the `full` image.

**Comment by danquack** [comment]
@CamJN curious if the design strategy is to either have a tag with `amd64` or if you plan on building multi arch under the version tag?

**Comment by CamJN** [comment]
> @CamJN can you provide an update on this? I see you published `phusion/passenger-full:2.3.1-arm64`. Is this something in beta at this point? Will support be added for all language specific images be included too? I see in your comment above that you're going to remove the architecture suffix and publish under one tag. Any timing on that?

I published arm versions of all the images, yes they are a β, I'll publish the next version of the images such that you don't need to specify the arch suffix, but before that happens I will wait for feedback on how the β images work.

**Comment by jwall-ys** [comment]
Hello all. Thanks for the work on all this the past few weeks. It's been pretty exciting to see the progress. I'm wondering, once all is said and done, how many of the images will have arm images? I ask because I'm specifically looking for an arm version of passenger-ruby25 and wondering if that will get the arm treatment once all is rolled out and the arm tags are folded into the regular tags.

Thanks!

**Comment by CamJN** [comment]
All of the maintained images will have arm versions once this is done, ruby 2.5 is EOL so we do not update the passenger-ruby25 image any more.

**Comment by danquack** [comment]
@CamJN this may be out of scope of this issue, so if there's a better place to ask Im happy to run it down there, but does the new arm build support `libnginx-mod-http-passenger`? I didn't see any arm install of that in the [oss binaries](https://oss-binaries.phusionpassenger.com/apt/passenger).

**Comment by CamJN** [comment]
@danquack the capacity to build that package has been added, they haven't been published to the repo yet as that will only happen in the next release when we have good feedback from this β. You could extract the binaries from the docker image I suppose for now.

**Comment by taleh007** [comment]
Sorry for long delay, we start testing images on M1 machine today. I will post results on Friday here

**Comment by taleh007** [comment]
I did couple of tests with Node, Ruby images on M1 and all works as it should for me.

If I can do anything to finish that work - I will be happy to help.

**Comment by taleh007** [comment]
@CamJN Hey, any possibility to see new version with both platforms during next week?

**Comment by taleh007** [comment]
@CamJN  Do you have any plans to finish this before end of year? Because for us it look almost ready

**Comment by CamJN** [comment]
@taleh007 plans, yes. Whether they work out depends on if I get some additional work done on our end for this month's Passenger release.

**Comment by CamJN** [comment]
Currently fighting with this: https://archlinuxarm.org/forum/viewtopic.php?f=57&t=14466&sid=f8664def5161aaac9b4c99f8fc5625d8&start=10

**Comment by CamJN** [comment]
per https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1895456 i might need to upgrade our CI server to jammy from focal.

**Comment by taleh007** [comment]
> Currently fighting with this: https://archlinuxarm.org/forum/viewtopic.php?f=57&t=14466&sid=f8664def5161aaac9b4c99f8fc5625d8&start=10

Do you need to run build on x64 machine?

**Comment by CamJN** [comment]
Yes, as we only have x86_64 machines.

**Comment by CamJN** [comment]
Now dealing with: https://github.com/aptly-dev/aptly/issues/243

**Comment by CamJN** [comment]
6.0.16 is released with arm packages, which means I can finally do this fo real. I've started building the new images, it takes about 10 hours on my laptop, so some time later today or tomorrow there should be arm images pushed without the need to specify the `-arm64` suffix.

**Comment by taleh007** [comment]
That is great news! Thanks a lot for your work :)

**Comment by CamJN** [comment]
done, thanks to everyone and especially @taleh007
